network/p2p/tracer/gossipSubMeshTracer.go

```diff
@@ -0,0 +1,184 @@
+package tracer
+
+import (
+	"sync"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/network/p2p"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+const (
+	// MeshLogIntervalMsg is the message logged by the tracer every logInterval.
+	MeshLogIntervalMsg = "topic mesh peers of local node since last heartbeat"
+)
+
+// GossipSubMeshTracer is a tracer that tracks the local mesh peers for each topic.
+// It also logs the mesh peers and updates the local mesh size metric.
+type GossipSubMeshTracer struct {
+	component.Component
+	pubsub.RawTracer
+
+	topicMeshMu    sync.RWMutex                    // to protect topicMeshMap
+	topicMeshMap   map[string]map[peer.ID]struct{} // map of local mesh peers by topic.
+	logger         zerolog.Logger
+	idProvider     module.IdentityProvider
+	loggerInterval time.Duration
+	metrics        module.GossipSubLocalMeshMetrics
+}
+
+var _ p2p.PubSubTracer = (*GossipSubMeshTracer)(nil)
+
+func NewGossipSubMeshTracer(
+	logger zerolog.Logger,
+	metrics module.GossipSubLocalMeshMetrics,
+	idProvider module.IdentityProvider,
+	loggerInterval time.Duration) *GossipSubMeshTracer {
+
+	g := &GossipSubMeshTracer{
+		RawTracer:      NewGossipSubNoopTracer(),
+		topicMeshMap:   make(map[string]map[peer.ID]struct{}),
+		idProvider:     idProvider,
+		metrics:        metrics,
+		logger:         logger.With().Str("component", "gossip_sub_topology_tracer").Logger(),
+		loggerInterval: loggerInterval,
+	}
+
+	g.Component = component.NewComponentManagerBuilder().
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			ready()
+			g.logLoop(ctx)
+		}).
+		Build()
+
+	return g
+}
+
+// GetMeshPeers returns the local mesh peers for the given topic.
+func (t *GossipSubMeshTracer) GetMeshPeers(topic string) []peer.ID {
+	t.topicMeshMu.RLock()
+	defer t.topicMeshMu.RUnlock()
+
+	peers := make([]peer.ID, 0, len(t.topicMeshMap[topic]))
+	for p := range t.topicMeshMap[topic] {
+		peers = append(peers, p)
+	}
+	return peers
+}
+
+// Graft is called when a peer is added to a topic mesh. The tracer uses this to track the mesh peers.
+func (t *GossipSubMeshTracer) Graft(p peer.ID, topic string) {
+	t.topicMeshMu.Lock()
+	defer t.topicMeshMu.Unlock()
+
+	lg := t.logger.With().Str("topic", topic).Str("peer_id", p.String()).Logger()
+
+	if _, ok := t.topicMeshMap[topic]; !ok {
+		t.topicMeshMap[topic] = make(map[peer.ID]struct{})
+	}
+	t.topicMeshMap[topic][p] = struct{}{}
+
+	id, exists := t.idProvider.ByPeerID(p)
+	if !exists {
+		lg.Warn().
+			Bool(logging.KeySuspicious, true).
+			Msg("grafted peer not found in identity provider")
+		return
+	}
+
+	lg.Info().Hex("flow_id", logging.ID(id.NodeID)).Str("role", id.Role.String()).Msg("grafted peer")
```

Comment:
how often is this called? will this be too noisy at `Info` level?

Reply:
The logging does not deem frequent. It is done once the startup of that node builds up its mesh and occasional changes in the local meshes of the nodes due to the peer scoring that it PRUNEs from some nodes and GRAFTs to others. Noting that the local mesh of a node cannot go beyond 12 in size so at most we have 12 * (number of channels) logs at the startup. 

This can be noisy only when a malicious peer starts spamming with GRAFTs and PRUNEs, which is detected and mitigated by our spamming prevention mechanism. Hence, overall, the impact is minimal. However, since our periodic logging intervals are designed to be long (to avoid being noisy), I think having such event-based logging upon GRAFT and PRUNEs can provide more scrutiny into what is happening between two intervals, e.g., some peers may GRAFT and PRUNE in shorter intervals than our logging interval so they less frequently are being logged. 

In conclusion, I am fine with switching to a lower-level log, but as far as I can see, it is less likely to be noisy unless a bug.